src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageItemListViewModel.cs

```
@@ -369,7 +395,69 @@ public bool IsPackageDeprecated
             }
         }
 
-        public Uri IconUrl { get; set; }
+        private Uri _iconUrl;
+        public Uri IconUrl
+        {
+            get { return _iconUrl; }
+            set
+            {
+                _iconUrl = value;
+                OnPropertyChanged(nameof(IconUrl));
+            }
+        }
+
+        private IconBitmapStatus _bitmapStatus;
+
+        public IconBitmapStatus BitmapStatus
+        {
+            get { return _bitmapStatus; }
+            set
+            {
+                if (_bitmapStatus != value)
+                {
+                    _bitmapStatus = value;
+                    OnPropertyChanged(nameof(BitmapStatus));
+                }
+            }
+        }
+
+        private BitmapSource _iconBitmap;
+        public BitmapSource IconBitmap
+        {
+            get
+            {
+                if (_iconBitmap == null)
+                {
+                    if (BitmapStatus == IconBitmapStatus.None)
+                    {
+                        (BitmapSource iconBitmap, IconBitmapStatus nextStatus) = GetInitialIconBitmapAndStatus();
+
+                        BitmapStatus = nextStatus;
+                        if (BitmapStatus == IconBitmapStatus.NeedToFetch)
+                        {
+                            BitmapStatus = IconBitmapStatus.Fetching;
+                            NuGetUIThreadHelper.JoinableTaskFactory
+                                .RunAsync(FetchThenSetIconBitmapAsync)
```

If there's a PostOnFailure, won't the status be `Fetching` forever?

---

yes, but at this point, that isn't really bad. just keeps showing the default icon forever.  we used to catch all exceptions which was bad practice. i've now added the ~8 exceptions we might get from icon decoding per a nice blog post: https://faithlife.codes/blog/2010/07/exceptions_thrown_by_bitmapimage_and_bitmapframe/

i propose we leave it Fetching for every. but we can discuss.

---

closing this conversation since you started 2 different ones.